martin/src/styles/mod.rs

```diff
@@ -0,0 +1,319 @@
+pub struct CatalogStyleEntry {
+    pub path: PathBuf,
+}
+
+pub type StyleCatalog = DashMap<String, CatalogStyleEntry>;
+
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct StyleConfig {
+    #[serde(flatten)]
+    pub unrecognized: UnrecognizedValues,
+}
+
+impl ConfigExtras for StyleConfig {
+    fn get_unrecognized(&self) -> &UnrecognizedValues {
+        &self.unrecognized
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+#[cfg_attr(test, serde_with::skip_serializing_none, derive(serde::Serialize))]
+pub struct StyleSources(DashMap<String, StyleSource>);
+
+#[derive(Clone, Debug)]
+#[cfg_attr(test, serde_with::skip_serializing_none, derive(serde::Serialize))]
+pub struct StyleSource {
+    path: PathBuf,
+}
+
+impl StyleSources {
+    pub fn resolve(config: &mut FileConfigEnum<StyleConfig>) -> FileResult<Self> {
+        let Some(cfg) = config.extract_file_config(None)? else {
+            return Ok(Self::default());
+        };
+
+        let mut results = Self::default();
+        let mut configs = BTreeMap::new();
+
+        if let Some(sources) = cfg.sources {
+            for (id, source) in sources {
+                configs.insert(id.clone(), source.clone());
+                results.add_sources(&id, &source.abs_path()?);
+            }
+        };
+
+        let mut paths_with_names = Vec::new();
+        for path in cfg.paths {
+            let Some(name) = path.file_name() else {
+                warn!("Ignoring style source with no name from {path:?}");
+                continue;
+            };
+            paths_with_names.push(path.clone());
+            results.add_sources(name.to_string_lossy().as_ref(), &path);
+        }
+
+        *config = FileConfigEnum::new_extended(paths_with_names, configs, cfg.custom);
+
+        Ok(results)
+    }
+
+    /// retrieve a styles' `PathBuf` from the internal catalog
+    pub fn style_json_path(&self, style_id: &str) -> Option<PathBuf> {
+        let style_id = style_id.trim_end_matches(".json").trim();
+        let item = self.0.get(style_id)?;
+        Some(item.path.clone())
+    }
+
+    /// an external representation of the internal catalog
+    #[must_use]
+    pub fn get_catalog(&self) -> StyleCatalog {
+        let entries = StyleCatalog::new();
+        for source in &self.0 {
+            entries.insert(
+                source.key().clone(),
+                CatalogStyleEntry {
+                    path: source.path.clone(),
+                },
+            );
+        }
+        entries
+    }
+
+    /// Adds all the contained files in the given file/directory as style sources.
+    fn add_sources(&mut self, id: &str, base_path: &PathBuf) {
+        match list_contained_files(base_path) {
+            Ok(contained_paths) => {
+                for child_path in contained_paths {
+                    let Some(name) = child_path.file_name() else {
+                        // FIXME: we should never panic in production code
```

Comment:
we should never crash on loading - asserts are evil

Reply:
this assertion is here, as it logically cannot happen.

The base path is checked if it is a valid file name one level up (where the id is extracted).

Reply:
wait no: actually you are entirely right:
if I `foo: bar.json` in the `sources`-part, the ID is not drawn from the parents filename

Reply:
@CommanderStorm I don't think this is the case - `id` is being ignored in the `fn add_sources` except to report warnings. Instead, it should be used when appropriate.